Testnet/NFTAuctionStore/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,199 @@
+﻿using Stratis.SmartContracts;
+
+[Deploy]
+public class NFTAuctionStore : SmartContract
+{
+    private void SetAuctionInfo(Address contract, ulong tokenId, AuctionInfo auctionInfo)
+    {
+        State.SetStruct($"AuctionInfo:{contract}:{tokenId}", auctionInfo);
+    }
+
+    public AuctionInfo GetAuctionInfo(Address contract, ulong tokenId)
+    {
+        return State.GetStruct<AuctionInfo>($"AuctionInfo:{contract}:{tokenId}");
+    }
+
+    public ulong GetRefund(Address address)
+    {
+        return State.GetUInt64($"Refund:{address}");
+    }
+
+    private void SetRefund(Address address, ulong balance)
+    {
+        State.SetUInt64($"Refund:{address}", balance);
+    }
+
+    public NFTAuctionStore(ISmartContractState state)
+        : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    public void Auction(Address contract, ulong tokenId, ulong startingPrice, ulong duration)
+    {
+        EnsureNotPayable();
+
+        var tokenOwner = GetOwner(contract, tokenId);
+
+        Assert(tokenOwner != Address, "The token is already on sale.");
+
+        EnsureCallerCanOperate(contract, tokenOwner);
+
+        TransferToken(contract, tokenId, tokenOwner, Address);
+
+        var auction = new AuctionInfo
+        {
+            Seller = tokenOwner,
+            EndBlock = checked(Block.Number + duration),
+            StartingPrice = startingPrice,
+            Ended = false
+        };
+
+        SetAuctionInfo(contract, tokenId, auction);
+        Log(new AuctionStartedLog { Contract = contract, TokenId = tokenId, EndBlock = auction.EndBlock, Seller = auction.Seller, startingPrice = auction.StartingPrice });
+    }
+
+    public void Bid(Address contract, ulong tokenId)
+    {
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(Block.Number < auction.EndBlock, "Auction ended.");
+
+        Assert(Message.Value > auction.HighestBid && Message.Value >= auction.StartingPrice, "The amount is not higher than highest bidder or starting price.");
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder);
+
+            SetRefund(auction.HighestBidder, balance + auction.HighestBid);
+        }
```

Comment:
Would be cleaner if this just refunded the previous highest bidder immediately.

Reply:
If previous highest bidder is a contract and we send back via Transfer method, the contract can reject incoming transfer in Receive method. So in this case, nobody would be able to Bid for the token. 

This is reason why withdraw pattern is preferred.